keps/sig-auth/20200507-external-tls-certificate-authenticator/README.md

```diff
@@ -0,0 +1,749 @@
+  [YubiKey](https://www.yubico.com/)
+- kubectl has no access to client key data
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP?  Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+- HSM support on the server-side, i.e., kubernetes-apiserver (although a
+  follow-up enhancement would be cool!)
+- Improving PKCS#11 support in the Go runtime or in a Go library
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what you're
+proposing, but should not include things like API designs or implementation.
+The "Design Details" section below is for the real nitty-gritty.
+-->
+
+This KEP proposes introducing a new authentication provider to enable a secure
+and complient usage of TLS client certificates within kubectl/client-go, by
+delegating digital key operations to extarnal processes, for example HSMs. The
+proposed authentication provider is similar to the [credentials
+plugin](https://kubernetes.io/docs/reference/access-authn-authz/authentication/#client-go-credential-plugins)
+in that it also uses an external process during a part of the authentication
+sequence, but contrary to the credentials plugin, it performs the sign operation
+externally, and returns only the product of it - the signature.
+
+In order to achieve a general solution, which can support various HSMs with
+different protocols, this KEP proposes to split the authenticator into two
+components and establish an API for the communication between them. The internal
+component provides only the general primitives for the authentication process,
+while leaving the implementation of a particular HSM protocol (for example
+PKCS#11) out of kubectl/client-go. Support of a specific protocol can be enabled
+by providing an external plugin that implements the proposed API.
+
+### User Stories
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system.  The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+As a developer or an operator, I want to be able to authenticate my API requests
+using a client certificate without a need of providing direct access to my
+private key data so that I can improve compliance and security of the whole
+system.
+
+To authenticate against the API:
+
+- The user issues a `kubectl` command, for example `kubectl get pods`.
+- Authentication provider calls the external plugin to obtain a client
+  certificate and signature.
+- External plugin prompts the user for a PIN to perform signing operation (if
+  the PIN is not provided in the configuration file).
+- External plugin returns a client certificate and a signature to client-go via
+  the authentication provider.
+- API server verifies the signature and processes the request.
```

Comment:
during the presentation of the SIG Auth meeting, and on https://github.com/kubernetes/kubernetes/issues/64783 there was an important question around caching of certificates. is the external provider going to cache/re-use the generated cert/key, for how long or is it going to generate new ones per every API request?

Reply:
I believe the caching was around TLS session keys.  We would have to prevent `kubectl` from doing that somehow.